File: 2025/03 March/28 Activity Selection/main.cpp
// Link: https://www.geeksforgeeks.org/problems/activity-selection-1587115620/1

#include <iostream>
#include <algorithm>
#include <map>
#include <vector>
using namespace std;

class Solution
{
public:
    int activitySelection(vector<int> &start, vector<int> &finish)
    {
        vector<pair<int, int>> activities;

        for (size_t i = 0; i < start.size(); i++)
        {
            activities.push_back({start[i], finish[i]});
        }

        sort(activities.begin(), activities.end(), [](const pair<int, int> &a, const pair<int, int> &b)
             { return a.second < b.second; });

        int count = 1;
        int lastFinishTime = activities[0].second;

        for (size_t i = 1; i < activities.size(); i++)
        {
            if (activities[i].first > lastFinishTime)
            {
                count++;
                lastFinishTime = activities[i].second;
            }
        }

        return count;
    }
};

int main()
{
    Solution obj;
    vector<int> start = {1, 3, 0, 5, 8, 5};
    vector<int> end = {2, 4, 6, 7, 9, 9};
    int result = obj.activitySelection(start, end);
    cout << "Maximum number of activities: " << result << endl;

    start = {10, 12, 20};
    end = {20, 25, 30};
    result = obj.activitySelection(start, end);
    cout << "Maximum number of activities: " << result << endl;

    return 0;
}

File: 2025/03 March/28 Activity Selection/main.py
# Link: https://www.geeksforgeeks.org/problems/activity-selection-1587115620/1


class Solution:
    def activitySelection(self, start, finish):
        res = a = []
        for i in range(len(start)):
            a = [start[i], finish[i]]
            res.append(a)
            a = []
        res.sort(key=lambda x: x[1])
        cnt = 1
        j = res[0][1]
        for i in range(1, len(res)):
            if j < res[i][0]:
                cnt += 1
                j = res[i][1]
        return cnt


if __name__ == "__main__":
    obj = Solution()
    start = [1, 3, 0, 5, 8, 5]
    finish = [2, 4, 6, 7, 9, 9]
    result = obj.activitySelection(start, finish)
    print(result)

    start = [10, 12, 20]
    finish = [20, 25, 30]
    result = obj.activitySelection(start, finish)
    print(result)
